base/reducedim.jl

```diff
@@ -301,14 +301,17 @@ function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted)
     keep, Idefault = Broadcast.shapeindexer(indsRt)
     if reducedim1(R, A)
         # keep the accumulator as a local variable when reducing along the first dimension
+        Eltype = _mapped_eltype(f, A)
+        Eltype = Eltype === eltype(R) ? Eltype : Any
+        pre, op_fast, post = _makefast_reduction(op, Eltype)
         i1 = first(axes1(R))
         @inbounds for IA in CartesianIndices(indsAt)
             IR = Broadcast.newindex(IA, keep, Idefault)
-            r = R[i1,IR]
+            r = pre(R[i1,IR])
             @simd for i in axes(A, 1)
-                r = op(r, f(A[i, IA]))
+                r = op_fast(r, pre(f(A[i, IA])))
             end
-            R[i1,IR] = r
+            R[i1,IR] = post(r)
         end
     else
         @inbounds for IA in CartesianIndices(indsAt)
```

Comment:
Shouldn't it be possible to apply the same series of transformations when reducing along non-first dimensions? Or am I missing where that has already been handled?

Reply:
Yes, exactly. And all the optimizations here are `reinterpret`-able.

Reply:
I dont think non-dim1 cases would be accelerated a lot as we will have to add a `map!(pre, dest, dest)` berfore the kernal loop, and a `map!(post, dest, dest)` after it.
And non dim1 minimum/maximum is vectorlized quite well
```julia
julia> @btime minimum($A, dims = 1);
  65.400 μs (7 allocations: 4.33 KiB)

julia> @btime minimum($A, dims = 2);
  56.500 μs (7 allocations: 4.33 KiB)
```
On my PC it's faster than dim1 with optimizaiton.

Reply:
A simple bench
```julia
f(b, c) = begin
    pre, op, post = Base._makefast_reducution(min, Float64)
    Base.initarray!(b, identity, min, true, c)
    map!(pre, b, b)
    Base.mapreducedim!(pre, op, b, c);
    map!(post, b, b)
end

julia> @btime f($B, $A);
  58.500 μs (0 allocations: 0 bytes)

julia> @btime minimum!($B, $A);
  57.700 μs (0 allocations: 0 bytes)
```